Source/SwiftLintFramework/Rules/Idiomatic/DuplicateImportsRule.swift

```diff
@@ -65,34 +120,71 @@ public struct DuplicateImportsRule: ConfigurationProviderRule, AutomaticTestable
             return lines[line - 1]
         }
 
-        var violations = [StyleViolation]()
-
-        for indexI in 0..<importLines.count {
-            for indexJ in indexI + 1..<importLines.count {
-                let firstLine = importLines[indexI]
-                let secondLine = importLines[indexJ]
-
-                guard firstLine.areImportsDuplicated(with: secondLine)
-                    else { continue }
-
-                let lineWithDuplicatedImport: Line = {
-                    if firstLine.importIdentifier?.count ?? 0 <= secondLine.importIdentifier?.count ?? 0 {
-                        return secondLine
-                    } else {
-                        return firstLine
-                    }
-                }()
-
-                let location = Location(file: file, characterOffset: lineWithDuplicatedImport.range.location)
-                let violation = StyleViolation(ruleDescription: Self.description,
-                                               severity: configuration.severity,
-                                               location: location)
-                violations.append(violation)
-            }
+        let importLineSlices = buildImportLineSlicesByImportSubpath(importLines: importLines)
+
+        let duplicateImports = findDuplicateImports(
+            file: file,
+            importLineSlicesGroupedBySubpath: Array(importLineSlices.values)
+        )
+
+        return duplicateImports.sorted(by: {
+            $0.range.lowerBound < $1.range.lowerBound
+        })
+    }
+
+    public func validate(file: SwiftLintFile) -> [StyleViolation] {
+        return duplicateImports(file: file).map { duplicateImport in
+            StyleViolation(
+                ruleDescription: Self.description,
+                severity: configuration.severity,
+                location: duplicateImport.location
+            )
+        }
+    }
+
+    public func correct(file: SwiftLintFile) -> [Correction] {
+        var contents = file.stringView.nsString
+        var corrections = [Correction]()
+
+        duplicateImports(file: file).reversed().filter {
+            file.ruleEnabled(violatingRange: $0.range, for: self) != nil
+        }.forEach { duplicateImport in
+            contents = contents.replacingCharacters(
+                in: duplicateImport.range,
+                with: ""
+            ).bridge()
+            corrections.append(
+                Correction(
+                    ruleDescription: Self.description,
+                    location: duplicateImport.location
+                )
+            )
```

Comment:
We could be a bit more explicit about what's happening here by separating the different steps from each other:
1. Collect all duplicated import for which the rule is enabled.
2. Map them to ranges and reduce the list by repeated execution of `contents.replacingCharacters(...)`.
3. Write the file.
4. Map the entries from the list created in the first step to locations, map them in turn to corrections and return the new list.

Reply:
Please see the revised implementation. I hope it is clearer now.

Reply:
It is. Thank you!